src/goto-programs/slice_global_inits.cpp

```diff
@@ -0,0 +1,143 @@
+/*******************************************************************\
+
+Module: Remove initializations of unused global variables
+
+Author: Daniel Poetzl
+
+Date:
+
+\*******************************************************************/
+
+#include <analyses/call_graph.h>
+
+#include <util/namespace.h>
+#include <util/std_expr.h>
+#include <util/cprover_prefix.h>
+#include <util/prefix.h>
+
+#include <goto-programs/goto_functions.h>
+
+#include "slice_global_inits.h"
+
+/*******************************************************************\
+
+Function: slice_global_inits
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+void slice_global_inits(
+  const namespacet &ns,
+  goto_functionst &goto_functions)
+{
+  // gather all functions reachable from the entry point
+
+  call_grapht call_graph(goto_functions);
+  const call_grapht::grapht &graph=call_graph.graph;
+
+  std::list<irep_idt> worklist;
+  std::set<irep_idt> functions_reached;
+
+  const irep_idt entry_point=goto_functionst::entry_point();
+
+  goto_functionst::function_mapt::const_iterator e_it;
+  e_it=goto_functions.function_map.find(entry_point);
+
+  if(e_it==goto_functions.function_map.end())
+    throw "entry point not found";
+
+  worklist.push_back(entry_point);
+
+  do
+  {
+    const irep_idt id=worklist.front();
+    worklist.pop_front();
+
+    functions_reached.insert(id);
+
+    const auto &p=graph.equal_range(id);
+
+    for(auto it=p.first; it!=p.second; it++)
+    {
+      const irep_idt callee=it->second;
+
+      if(functions_reached.find(callee)==functions_reached.end())
+        worklist.push_back(callee);
+    }
+  } while(!worklist.empty());
```

Comment:
Note to self: there is a patch to add functionality just listing all reachable functions, which has very similar code. I should file a pull request for that one.